### ISSUE TITLE
Tensor datum type error: tensor is F16, accessed as F32

### ISSUE THREAD

**Issue by Rorical** [sonos/tract#874]
I was trying to run a model exported from PyTorch with float16.
When processing the InstanceNormalization node, I got an error: 
```
Evaluating #147 "InstanceNormalization_7" InstanceNorm

Caused by:
    0: Evaluating #13 "adhoc.epsilon" AddUnary
    1: Tensor datum type error: tensor is F16, accessed as F32
```

The Inputs and Outputs of this node are both fp16, but its attribute `epsilon` must be fp32 according to its definition.

I couldn't figure it out, is it something related to the operations between fp16 and fp32 tensors?

**Comment by kali** [comment]
It's most certainly a bug. Float 16 support is relatively recent, and it has not been heavily tested with float16 models (juste using tract fp32->fp16 conversion utilities). Can you share the onnx model with me so I can have a look ?

**Comment by Rorical** [comment]
Yes, I uploaded it on my cloud: https://cloud.rorical.blue/s/DgCK
The test input size can be (1, 4, 64, 64)

Specifically, at this line
https://github.com/sonos/tract/blob/531ec1050cde2f523dda44e350d624a378df2228/onnx/src/ops/nn/instance_norm.rs#L71
it uses epsilon which is fp32 and vari which is fp16 in my model.
This error occurs at `eval_out_of_place` in the add operation, where `a.to_array_view::<$typ>()?;` wants to read fp16 `a` as `$typ` fp32.

**Comment by kali** [comment]
Thanks for sharing the model! Can you give a try to the branch I just linked ? Tell me how it goes.

**Comment by Rorical** [comment]
Great work! This error is solved.

But there is another error occurred:
```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\ops\nn\softmax\mod.rs:177:59
stack backtrace:
   0: std::panicking::begin_panic_handler
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library\std\src\panicking.rs:575
   1: core::panicking::panic_fmt
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library\core\src\panicking.rs:65
   2: core::panicking::panic
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535/library\core\src\panicking.rs:115
   3: enum2$<core::option::Option<core::cmp::Ordering> >::unwrap<core::cmp::Ordering>
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535\library\core\src\option.rs:778
   4: tract_core::ops::nn::softmax::softmax_inner::closure$0<half::binary16::f16,ndarray::dimension::dim::Dim<ndarray::dimension::dynindeximpl::IxDynImpl> >
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\ops\nn\softmax\mod.rs:177
   5: core::ops::function::impls::impl$4::call_once<tuple$<ref$<ref$<half::binary16::f16> >,ref$<ref$<half::binary16::f16> > >,tract_core::ops::nn::softmax::softmax_inner::closure_env$0<half::binary16::f16,ndarray::dimension::dim::Dim<ndarray::dimension::dynind
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535\library\core\src\ops\function.rs:309
   6: core::cmp::max_by<ref$<half::binary16::f16>,ref_mut$<tract_core::ops::nn::softmax::softmax_inner::closure_env$0<half::binary16::f16,ndarray::dimension::dim::Dim<ndarray::dimension::dynindeximpl::IxDynImpl> > > >
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535\library\core\src\cmp.rs:1317
   7: core::iter::traits::iterator::Iterator::max_by::fold::closure$0<ref$<half::binary16::f16>,tract_core::ops::nn::softmax::softmax_inner::closure_env$0<half::binary16::f16,ndarray::dimension::dim::Dim<ndarray::dimension::dynindeximpl::IxDynImpl> > >
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535\library\core\src\iter\traits\iterator.rs:3062
   8: core::iter::traits::iterator::Iterator::fold<core::slice::iter::Iter<half::binary16::f16>,ref$<half::binary16::f16>,core::iter::traits::iterator::Iterator::max_by::fold::closure_env$0<ref$<half::binary16::f16>,tract_core::ops::nn::softmax::softmax_inner::
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535\library\core\src\iter\traits\iterator.rs:2414
   9: ndarray::iterators::impl$13::fold<half::binary16::f16,ndarray::dimension::dim::Dim<ndarray::dimension::dynindeximpl::IxDynImpl>,ref$<half::binary16::f16>,core::iter::traits::iterator::Iterator::max_by::fold::closure_env$0<ref$<half::binary16::f16>,tract_c
             at .cargo\registry\src\github.com-1ecc6299db9ec823\ndarray-0.15.6\src\iterators\mod.rs:398
  10: core::iter::traits::iterator::Iterator::reduce<ndarray::iterators::Iter<half::binary16::f16,ndarray::dimension::dim::Dim<ndarray::dimension::dynindeximpl::IxDynImpl> >,core::iter::traits::iterator::Iterator::max_by::fold::closure_env$0<ref$<half::binary16
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535\library\core\src\iter\traits\iterator.rs:2450
  11: core::iter::traits::iterator::Iterator::max_by<ndarray::iterators::Iter<half::binary16::f16,ndarray::dimension::dim::Dim<ndarray::dimension::dynindeximpl::IxDynImpl> >,tract_core::ops::nn::softmax::softmax_inner::closure_env$0<half::binary16::f16,ndarray:
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535\library\core\src\iter\traits\iterator.rs:3065
  12: tract_core::ops::nn::softmax::softmax_inner<half::binary16::f16,ndarray::dimension::dim::Dim<ndarray::dimension::dynindeximpl::IxDynImpl> >
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\ops\nn\softmax\mod.rs:177
  13: tract_core::ops::nn::softmax::Softmax::eval_t<half::binary16::f16>
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\ops\nn\softmax\mod.rs:138
  14: tract_core::ops::nn::softmax::impl$2::eval
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\ops\nn\softmax\mod.rs:108
  15: tract_core::model::typed::impl$0::wire_node::closure$0<str,tract_core::ops::nn::softmax::Softmax>
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\model\typed.rs:55
  16: tract_core::model::typed::impl$0::wire_node<str,tract_core::ops::nn::softmax::Softmax>
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\model\typed.rs:62
  17: tract_hir::ops::nn::softmax::impl$0::wire
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\hir\src\ops\nn\softmax.rs:55
  18: tract_hir::ops::expandable::impl$2::eval
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\hir\src\ops\expandable.rs:86
  19: tract_core::plan::eval<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn$<tract_hir::infer::ops::InferenceOp>,alloc::alloc::Global> >
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\plan.rs:491
  20: core::ops::function::FnMut::call_mut<enum2$<core::result::Result<smallvec::SmallVec<array$<alloc::sync::Arc<tract_data::tensor::Tensor>,4> >,anyhow::Error> > (*)(ref_mut$<tract_core::plan::SessionState>,enum2$<core::option::Option<ref_mut$<dyn$<tract_core
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535\library\core\src\ops\function.rs:166
  21: tract_core::plan::SimpleState<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn$<tract_hir::infer::ops::InferenceOp>,alloc::alloc::Global>,tract_core::model::graph::Graph<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn$<tract_hir::in
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\plan.rs:264
  22: tract_core::plan::SimpleState<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn$<tract_hir::infer::ops::InferenceOp>,alloc::alloc::Global>,tract_core::model::graph::Graph<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn$<tract_hir::in
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\plan.rs:199
  23: tract_core::plan::SimpleState<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn$<tract_hir::infer::ops::InferenceOp>,alloc::alloc::Global>,tract_core::model::graph::Graph<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn$<tract_hir::in
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\plan.rs:177
  24: tract_core::plan::SimplePlan<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn$<tract_hir::infer::ops::InferenceOp>,alloc::alloc::Global>,tract_core::model::graph::Graph<tract_hir::infer::fact::InferenceFact,alloc::boxed::Box<dyn$<tract_hir::inf
             at .cargo\git\checkouts\tract-01a3138c7a4b58a3\aac4f0c\core\src\plan.rs:115
  25: StableOnnx::main
             at .\src\main.rs:11
  26: core::ops::function::FnOnce::call_once<void (*)(),tuple$<> >
             at /rustc/0da281b6068a7d889ae89a9bd8991284cc9b7535\library\core\src\ops\function.rs:251
```

Probably line
https://github.com/sonos/tract/blob/main/core/src/ops/nn/softmax/mod.rs#L177
got a None value from `partial_cmp`, since it is called on an fp16 value.

**Comment by kali** [comment]
Hello, I was unable to reproduce the bug, but pushed a possible fix on the same branch. Crossing my fingers hoping it's the last one :)

**Comment by Rorical** [comment]
Yeah, that might be the last one, thanks for your excellent work!
